File: LeetCode/Python/#563 Binary Tree Tile.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findTilt(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.res = 0
        def sums(roots):
            if not roots:
                return 0
            left,right = sums(roots.left),sums(roots.right)
            self.res+=abs(left-right)
            return roots.val+left+right
        sums(root)
        return self.res

File: LeetCode/JavaScript/#563 Binary Tree Tile.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
/*var findTilt = function(root) {
	var res = 0;
	var temp;
    if(root===null){
    	return 0;
    }
    if(root.left&&root.right){
    	temp = Math.abs(root.left.val-root.right.val);
    	res = res+temp+findTilt(root.left)+findTilt(root.right);
    }else if(root.left){
    	temp = root.left.val;
    	res = res+temp+findTilt(root.left);
    }else if(root.right){
    	temp = root.right.val;
    	res = res+temp+findTilt(root.right);
    }
    return res;
};
*/

var findTilt = function(root) {
	var res = 0;
	function sum(root){
		
		if(root===null){
	    	return 0;
	    }
	    let left = sum(root.left);
	    let right = sum(root.right);
		res +=Math.abs(left-right);
		return root.val+left+right;
	}
    sum(root);
    return res;
};
